provisioning/transport/amqp/src/AmqpClientConnection.cs

```diff
@@ -153,5 +175,57 @@ void OnConnectionClosed(object o, EventArgs args)
 
             return websocket;
         }
+
+        void OnWriteHeaderComplete(TransportAsyncCallbackArgs args)
+        {
+            if (args.Exception != null)
+            {
+                this.Complete(args);
+                return;
+            }
+
+            byte[] headerBuffer = new byte[AmqpConstants.ProtocolHeaderSize];
+            args.SetBuffer(headerBuffer, 0, headerBuffer.Length);
+            args.CompletedCallback = this.OnReadHeaderComplete;
+            args.Transport.ReadAsync(args);
+        }
+
+        void OnReadHeaderComplete(TransportAsyncCallbackArgs args)
+        {
+            if (args.Exception != null)
+            {
+                this.Complete(args);
+                return;
+            }
+
+            try
+            {
+                ProtocolHeader receivedHeader = new ProtocolHeader();
+                receivedHeader.Decode(new ByteBuffer(args.Buffer, args.Offset, args.Count));
+                if (!receivedHeader.Equals(this.sentHeader))
+                {
+                    throw new AmqpException(AmqpErrorCode.NotImplemented, $"The requested protocol version {this.sentHeader} is not supported. The supported version is {receivedHeader}");
+                }
+
+                SaslTransportProvider provider = this._amqpSettings.GetTransportProvider<SaslTransportProvider>();
+                var transport = provider.CreateTransport(args.Transport, true);
+                tcs.TrySetResult(transport);
+            }
+            catch (Exception ex)
+            {
+                args.Exception = ex;
+                this.Complete(args);
+            }
+        }
+
+        void Complete(TransportAsyncCallbackArgs args)
+        {
+            if (args.Exception != null && args.Transport != null)
+            {
+                args.Transport.SafeClose(args.Exception);
+                args.Transport = null;
+                tcs.TrySetException(args.Exception);
```

Comment:
Why is `Complete` only performing Task completion for exceptions and not the normal case?
Is this incorrectly named? (Maybe `CompleteWithError`?)